src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -895,6 +895,510 @@ uct_ib_mlx5_devx_dereg_keys(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh)
+uct_ib_mlx5_devx_umr_mkey_bind(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mr_t *mr,
+                               uct_ib_mlx5_devx_umr_t **umr)
+{
+    struct ibv_mr *ib_mr  = mr->super.ib;
+    struct ibv_mw mw      = {
+        .rkey = 0,
+        .type = IBV_MW_TYPE_1
+    };
+    struct ibv_send_wr wr = {
+        .wr_id      = 0,
+        .next       = NULL,
+        .opcode     = IBV_WR_BIND_MW,
+        .send_flags = IBV_SEND_INLINE | IBV_SEND_SIGNALED,
+        .bind_mw    = {
+            .mw        = &mw,
+            .rkey      = 0,
+            .bind_info = {
+                .mr              = ib_mr,
+                .length          = ib_mr->length,
+                .addr            = (uint64_t)ib_mr->addr,
+                .mw_access_flags = UCT_IB_MLX5_MD_UMEM_ACCESS
+            }
+        }
+    };
+    ucs_status_t status;
+
+    *umr = uct_ib_mlx5_devx_umr_mkey_pool_get(md);
+    if (NULL == *umr) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    /* Set UMR lkey to be registered with given MR */
+    mw.rkey         = (*umr)->umr_mkey->lkey;
+    wr.bind_mw.rkey = (*umr)->umr_mkey->lkey;
+
+    status = uct_ib_mlx5_devx_umr_post_sync(md, &wr, *umr, "UMR registration");
+    if (status != UCS_OK) {
+        uct_ib_mlx5_devx_umr_mkey_invalidate(md, *umr);
+        *umr = NULL;
+        return status;
+    }
+
+    ucs_trace("%s: registered addr %p length %zu lkey 0x%x to "
+              UCT_IB_MLX5_UMR_FMT, uct_ib_device_name(&md->super.dev),
+              ib_mr->addr, ib_mr->length, ib_mr->lkey,
+              UCT_IB_MLX5_UMR_ARG(*umr));
+
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE uint64_t
+uct_ib_mlx5_devx_umr_mkey_hash_code(const uct_ib_md_packed_mkey_t *packed_mkey)
+{
+    return (uint64_t)packed_mkey->vhca_id << 32 | packed_mkey->lkey;
+}
+
+/*
+ * Indicate that given mkey is an UMR mkey and XGVMI enabled.
+ */
+static UCS_F_ALWAYS_INLINE int uct_ib_mlx5_devx_umr_mkey_is_xgvmi(uint32_t lkey)
+{
+    return (UCT_IB_MLX5_MKEY_TAG_UMR == (lkey & 0xFF));
+}
+
+/*
+ * Mark UMR mkey as XGVMI enabled by setting mkey tag (8 LSB in mkey).
+ * This serves for 2 purposes:
+ *  1. On the exporter side - to identify that the mkey is XGVMI enabled, and
+ *     therefore avoiding repetitive XGVMI enablement.
+ *  2. On the importer side - to identify that received rkey is UMR based, and
+ *     therefore should be looked up and stored in UMR alias hash map
+ */
+static UCS_F_ALWAYS_INLINE uint32_t
+uct_ib_mlx5_devx_umr_mkey_set_xgvmi(uct_ib_mlx5_devx_umr_t *umr)
+{
+    umr->umr_mkey->lkey |= UCT_IB_MLX5_MKEY_TAG_UMR;
+    umr->umr_mkey->rkey |= UCT_IB_MLX5_MKEY_TAG_UMR;
+    return umr->umr_mkey->lkey;
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_umr_mkey_hash_find(uct_ib_mlx5_md_t *md,
+                                    const uct_ib_md_packed_mkey_t *packed_mkey,
+                                    uct_ib_mlx5_devx_umr_alias_t *umr_alias)
+{
+    uint64_t input_key = uct_ib_mlx5_devx_umr_mkey_hash_code(packed_mkey);
```

Comment:
IMO it's a better usage to define umr_mkey_map hash table as uct_ib_md_packed_mkey_t key type (instead of uint64_t), instead of calling uct_ib_mlx5_devx_umr_mkey_hash_code directly outside the hash

Reply:
Yes, it looks like a good idea.
I implemented it in a test project, and I don't really like the outcome.
```
static inline uint64_t
uct_ib_mlx5_devx_umr_mkey_hash_code2(uct_ib_md_packed_mkey_t packed_mkey)
{
    return ((uint64_t)packed_mkey.vhca_id << 32) | packed_mkey.lkey;
}

static int uct_ib_mlx5_devx_umr_mkey_equal2(uct_ib_md_packed_mkey_t a,
                                            uct_ib_md_packed_mkey_t b)
{
    // return (a.vhca_id == b.vhca_id) && (a.lkey == b.lkey); <-- slow
    return memcmp(&a, &b, sizeof(a)) == 0; // is it always safe wrt padding?
}

KHASH_INIT(test_map2, uct_ib_md_packed_mkey_t, uint64_t, 1,
           uct_ib_mlx5_devx_umr_mkey_hash_code2, uct_ib_mlx5_devx_umr_mkey_equal2);
```
First, it's more verbose, cause you have to provide custom `hash_equals` function.
Then, lookup is a bit slower then with the original impl:
```
test_map1: 1138
test_map2: 1334
```
The reason is that equal function calls memcmp, which is slower then the plain integers comparison. So I would keep the original impl, which is shorter and faster.

FYI comparison test: /labhome/iyastrebov/ws/misc/hash.c